### PR TITLE
Add few missing operations and tests

### PR DESCRIPTION
**Pull request by alibazlamit**
Fixed an issue with the test that was removing all datacenters of once account.

minor changes to the copyright headers

@edevenport

### CODE REVIEW

File: src/main/groovy/SDKTester.groovy [hunk 1]
```diff
@@ -15,312 +15,387 @@ class SDKTester {
...
+                v.all.each { id ->
+                    def _v = v.read(id)
+                    assert _v.delete()
+                    println "deleted $_v"
+                }
+
+                v = v.create()
+                println "recreated $v"
+                
+
```
Comment (edevenport) [inline src/main/groovy/SDKTester.groovy]: The extra newlines should be removed.

File: src/main/groovy/SDKTester.groovy [hunk 2]
```diff
@@ -15,312 +15,387 @@ class SDKTester {
...
+                println "selected image: $i"
+
+                // -------- image update ---------
+
+                i.name = 'my ubuntu'
+                i._public = false
+                try {
+                    i.update()
+                } catch (HttpResponseException e) {
+                    assert  e.statusCode == 422 // stoopid API
```
Comment (edevenport) [inline src/main/groovy/SDKTester.groovy]: I know the comment is from Frank, but we should remove it for release. It provides no details.

File: src/main/groovy/SDKTester.groovy [hunk 3]
```diff
@@ -15,312 +15,387 @@ class SDKTester {
...
+                Snapshot sn = new Snapshot()
+                sn.all.collect{sn.read(it) as Snapshot}.each{
+                    println "found $it"
+                    it.name="${it.name} updated"
+                    it.description="${it.description} updated"
+                    try {
+                        assert it.update()
+                        
+                    } catch (HttpResponseException e) {
+                        assert  e.statusCode == 422 // stoopid API
```
Comment (edevenport) [inline src/main/groovy/SDKTester.groovy]: Unnecessary comment for release.

File: src/main/groovy/SDKTester.groovy [hunk 4]
```diff
@@ -15,312 +15,387 @@ class SDKTester {
...
+                assert attach(s, i)
+                println "attached $i to $s"
+                assert attachedCDROMs(s).collect{it.id}.contains(i.id)
+                assert attachedCDROM(s,i).id.equals(i.id)
+
+                assert detach(s, i)
+                println "detached $i from $s"
+
+                assert stop(s)
+                assert (s = s.read()).vmState =~ /(?i)shutoff/ // wtf
```
Comment (edevenport) [inline src/main/groovy/SDKTester.groovy]: Again, another unnecessary comment and inappropriate for release.

File: src/main/groovy/com/profitbricks/sdk/Commands.groovy [hunk 5]
```diff
@@ -127,9 +148,9 @@ class Commands {
      */
     static final Snapshot snapshot(final Volume volume, final String name = '', final String description = '') {
         def req = requestFor("${volume.resource}/${volume.id}/create-snapshot") + [body: [
-            name: name ?: "snapshot_of_${volume.name}",
-            description: description ?: "source volume: ${volume.name} (${volume.id}), created: ${new Date().format('dd.MM.yyyy HH:mm')}"
-        ]]
+                name: name ?: "snapshot_of_${volume.name}",
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/Commands.groovy]: Too many white spaces here and below?

File: src/main/groovy/com/profitbricks/sdk/Commands.groovy [hunk 6]
```diff
@@ -168,6 +189,29 @@ class Commands {
         }
         throw new HttpResponseException(resp?.status as int, "NIC not associated")
     }
+    
+
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/Commands.groovy]: Remove a newline.

File: src/main/groovy/com/profitbricks/sdk/Commands.groovy [hunk 7]
```diff
@@ -168,6 +189,29 @@ class Commands {
...
+    /**
+     * gets an associated NIC to a loadbalancer
+     *
+     * @param LoadBalancer the loadbalancer with the associated nics
+     * @param nicId the nic id attached to the loadbalancer
+     * @return NIC associated
+     */
+    static final NIC associatedNic(final LoadBalancer loadBalancer,final String nicId) {
+        return new NIC().from(API.get(requestFor("${loadBalancer.resource}/${loadBalancer.id}/balancednics/$nicId"))?.data) as NIC
+
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/Commands.groovy]: Remove the newline.

File: src/main/groovy/com/profitbricks/sdk/Common.groovy [hunk 8]
```diff
@@ -12,90 +12,94 @@ import static groovyx.net.http.ContentType.JSON
  * some functions enclosing convenience functionality
  *
  * Created by fudge on 01/02/17.
- * (c)2015 Profitbricks.com
+ * Copyright (c) 2017, ProfitBricks GmbH
  */
 @Log4j2
 final class Common {
-    private final static RESTClient client = new RESTClient(URLParts.host)
+   private final static RESTClient client = new RESTClient(URLParts.host)
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/Common.groovy]: Much of this file is shifted from 4 space indention to 3 space indention. Should be adjusted back to 4 space indention to match the rest of the SDK.

File: src/main/groovy/com/profitbricks/sdk/model/Request.groovy [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    
+    static class Target {
+        String id
+        String type
+        String href
+    }
+    final String resource = 'requests'
+
+    //    @Override
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/model/Request.groovy]: Can we remove these comments if they are not necessary?

File: src/main/groovy/com/profitbricks/sdk/model/Request.groovy [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    //    final LAN read(final id = id) { (super.read(id) as LAN)?.with(dataCenter) }
+    //
+    @Override
+    final String getResource() { "/requests" }
+    //
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/model/Request.groovy]: Are these comments necessary?

File: src/main/groovy/com/profitbricks/sdk/Commands.groovy [hunk 11]
```diff
@@ -168,6 +189,29 @@ class Commands {
         }
         throw new HttpResponseException(resp?.status as int, "NIC not associated")
     }
+    
+    
+    /**
+     * gets the associated NICS to a loadbalancer
```
Comment (edevenport) [inline src/main/groovy/com/profitbricks/sdk/Commands.groovy]: NICS and nics should consistently be NICs. Also, "gets the associated NICs to a loadbalancer" should be "gets the associated NICs **of** a loadbalancer".

**Comment by alibazlamit** [comment]
Updated with requested changes

**Pull request merged**